File: BFS/게임맵최단거리.py
from collections import deque

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1] 

def BFS(maps, start):
    queue = deque()
    queue.append(start) #(x,y)
    n = len(maps) # y축
    m = len(maps[0]) # x축
    
    while queue:
        x, y = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
        
            if nx<0 or nx>m-1 or ny<0 or ny>n-1:
                continue
            if maps[ny][nx] == 1:
                queue.append((nx,ny))
                maps[ny][nx] = maps[y][x] + 1
            
    return maps[n-1][m-1]
        

def solution(maps):
    answer = BFS(maps, (0,0))
    if answer <= 1:
        answer = -1
    return answer

File: C++연습/게임맵최단거리.cpp
#include<vector>
#include<iostream>
#include<queue>
#include<algorithm>
using namespace std;

void bfs();
vector<vector<int>> board = {{}};
vector<vector<int>> visited = {{}};
vector<vector<int>> d = {{-1,0}, {0,1}, {1,0}, {0,-1}};
int INF = 123456789;

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    board = maps;
    visited = maps;
    for(int i=0; i<maps.size(); i++) {
        for(int j=0; j<maps[0].size(); j++) {
            board[i][j] = INF;
            visited[i][j] = 0;
        }
    }

    
    // bfs
    queue<pair<int, int>> q;
    q.push({0, 0});
    board[0][0] = maps[0][0];

    
    while (!q.empty()) {
        pair<int, int> now = q.front();
        q.pop();
        int x = now.first;
        int y = now.second;
        
        if (not visited[x][y]) {
            visited[x][y] = 1;
        
            for(int i=0; i<d.size(); i++){
                int nx = x + d[i][0];
                int ny = y + d[i][1];

                if (nx<0 || nx>maps.size()-1 || ny<0 || ny>maps[0].size()-1 || maps[nx][ny]==0){
                    continue;
                }

                board[nx][ny] = min(board[nx][ny], board[x][y]+maps[nx][ny]);
                q.push({nx, ny});
            }
        }
    }
    
    answer = board[maps.size()-1][maps[0].size()-1];
    if (answer == INF) {
        return -1;
    }
    
    return answer;
}

